monitoring/v1alpha1/types.go

```diff
@@ -0,0 +1,280 @@
+//
+// +k8s:openapi-gen=true
+type RuleGroup struct {
+	// name is the name of the group.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// interval is how often rules in the group are evaluated.  If not specified,
+	// it defaults to the global.evaluation_interval configured in Prometheus,
+	// which itself defaults to 1 minute.  This is represented as a Prometheus
+	// duration, for details on the format see:
+	// - https://prometheus.io/docs/prometheus/latest/configuration/configuration/#duration
+	//
+	// +kubebuilder:validation:Pattern:="((([0-9]+)y)?(([0-9]+)w)?(([0-9]+)d)?(([0-9]+)h)?(([0-9]+)m)?(([0-9]+)s)?(([0-9]+)ms)?|0)"
+	// +optional
+	Interval string `json:"interval,omitempty"`
+
+	// rules is a list of sequentially evaluated alerting rules.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	Rules []Rule `json:"rules"`
+}
+
+// Rule describes an alerting rule.
+// See Prometheus documentation:
+// - https://www.prometheus.io/docs/prometheus/latest/configuration/alerting_rules
+//
+// +k8s:openapi-gen=true
+type Rule struct {
+	// alert is the name of the alert. Must be a valid label value, i.e. only
+	// contain ASCII letters, numbers, and underscores.
+	//
+	// +kubebuilder:validation:Pattern:="^[a-zA-Z_][a-zA-Z0-9_]*$"
+	// +required
+	Alert string `json:"alert"`
+
+	// expr is the PromQL expression to evaluate. Every evaluation cycle this is
+	// evaluated at the current time, and all resultant time series become
+	// pending/firing alerts.
+	//
+	// +required
+	Expr intstr.IntOrString `json:"expr"`
+
+	// for is the time period after which alerts are considered firing after first
+	// returning results.  Alerts which have not yet fired for long enough are
+	// considered pending. This is represented as a Prometheus duration, for
+	// details on the format see:
+	// - https://prometheus.io/docs/prometheus/latest/configuration/configuration/#duration
+	//
+	// +kubebuilder:validation:Pattern:="((([0-9]+)y)?(([0-9]+)w)?(([0-9]+)d)?(([0-9]+)h)?(([0-9]+)m)?(([0-9]+)s)?(([0-9]+)ms)?|0)"
+	// +optional
+	For string `json:"for,omitempty"`
+
+	// labels to add or overwrite for each alert.
+	//
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// annotations to add to each alert.
+	//
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+// AlertingRuleStatus is the status of an AlertingRule resource.
+type AlertingRuleStatus struct {
+	// observedGeneration is the last generation change you've dealt with.
+	//
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// prometheusRule is the generated PrometheusRule for this AlertingRule.
+	//
+	// +optional
+	PrometheusRule PrometheusRuleRef `json:"prometheusRule,omitempty"`
+}
+
+// PrometheusRuleRef is a reference to an existing PrometheusRule object.
+type PrometheusRuleRef struct {
+	// name of the referent; More info: http://kubernetes.io/docs/user-guide/identifiers#names
```

Comment:
Given the three different names contained for each element of this API and the multiplicity of each, how does a single name here help?

What would it mean for this to be empty?

Reply:
> What all do you think we need here to point to the generated rule usefully?

So the confusion here is as to whether this reference refers to something inside K8s or inside Prometheus. My suggestion would be to just contain `Name`. This looks like an upstream LocalObjectReference to me, but we can add documentation to make sure it's obvious that this is referring to a Kubernetes Object in a particular namespace, of a certain group and kind.

Reply:
Yeah, sorry, this is a Kubernetes object. You're exactly right, it's basically a `LocalObjectReference`. Should I just use the `LocalObjectReference` type directly for this and add documentation then?

Reply:
We don't use the upstream types from metav1 for a few reasons, one is so that if they were to add new fields, we don't automatically gain those in our API, that would be bad, but also, so we can add relevant documentation to our copies, so I'd keep your struct and documentation, but just make it down to have the one Name field.

Previously I've added a comment like this at the end of the struct as hint for devs
```
type PrometheusRuleRef {
  ...

  // This is a struct so that we can support future expansion of fields within the reference should we ever need to.
}
```